wayland-scanner/src/common_gen.rs

```diff
@@ -626,11 +626,50 @@ pub(crate) fn gen_interface(
     low_name: &str,
     version: u32,
     addon: Option<TokenStream>,
+    side: Side,
 ) -> TokenStream {
+    let object_type = Ident::new(
+        match side {
+            Side::Client => "Proxy",
+            Side::Server => "Resource",
+        },
+        Span::call_site(),
+    );
+    let as_object_name = Ident::new(
+        match side {
+            Side::Client => "as_proxy",
+            Side::Server => "as_resource",
+        },
+        Span::call_site(),
+    );
+    let as_object_doc_attr = to_doc_attr(&format!("Returns the underlying `{}`.", object_type));
+
     let version_lit = Literal::u32_unsuffixed(version);
 
     quote! {
-        pub struct #name;
+        #[derive(Clone, Eq, PartialEq)]
+        pub struct #name(#object_type<#name>);
+
+        impl #name {
+            #as_object_doc_attr
+            pub fn #as_object_name(&self) -> &#object_type<Self> {
+                &self.0
+            }
+        }
```

Comment:
Maybe to avoid name clashes and to be more generic overall, this could be an implementation of the [`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html) trait ? So an `impl AsRef<Proxy<#name>> for #name { ... }`.

Reply:
Oh yeah, good idea.